sz.py

```diff
@@ -1,28 +1,70 @@
 #!/usr/bin/env python3
-import os
-from pathlib import Path
+import os, sys
 import token
 import tokenize
 import itertools
 from tabulate import tabulate
 
 TOKEN_WHITELIST = [token.OP, token.NAME, token.NUMBER, token.STRING]
 
-if __name__ == "__main__":
-  headers = ["Name", "Lines", "Tokens/Line"]
+def gen_stats(base_path="."):
   table = []
-  for path, subdirs, files in os.walk("tinygrad"):
+  for path, _, files in os.walk(os.path.join(base_path, "tinygrad")):
     for name in files:
       if not name.endswith(".py"): continue
-      filepath = Path(path) / name
+      filepath = os.path.join(path, name)
+      relfilepath = os.path.relpath(filepath, base_path)
       with tokenize.open(filepath) as file_:
         tokens = [t for t in tokenize.generate_tokens(file_.readline) if t.type in TOKEN_WHITELIST]
         token_count, line_count = len(tokens), len(set([t.start[0] for t in tokens]))
-        table.append([filepath.as_posix(), line_count, token_count/line_count])
+        table.append([relfilepath, line_count, token_count/line_count])
+  return table
+
+def gen_diff(table_old, table_new):
+  table = []
+  files_new = set([x[0] for x in table_new])
+  files_old = set([x[0] for x in table_old])
+  added, deleted, unchanged = files_new - files_old, files_old - files_new, files_new & files_old
+  if added:
+    for file in added:
+      for line in table_new:
+        if file in line: table.append([line[0], line[1], line[1] - 0, line[2], line[2]-0]) 
+  if deleted:
+    for file in deleted:
+      for line in table_old: 
+        if file in line: table.append([line[0], 0, 0 - line[1], 0, 0-line[2]])
+  if unchanged:
+    for file in unchanged:
+      for line in table_old: 
+        if file in line: line_old = line 
+      for line in table_new: 
+        if file in line: line_new = line
+      if line_new[1]-line_old[1] != 0 or line_new[2]-line_old[2] != 0: 
+        table.append([line_new[0], line_new[1], line_new[1]-line_old[1], line_new[2], line_new[2]-line_old[2]])
```

Comment:
you should be able to simplify this quite a bit with something along the lines of (not tested): 
``` python
for file in files_new.union(files_old):
  if (file in unchanged) and (table_old[file] != table_new[file]):
    table.append([table_new[file][0], table_new[file][1], table_new[file][1]-table_old[file][1], table_new[file][2], table_old[file][file][2]-line_old[2]])
  else:
    table.append([line[0], line[1], line[1] - 0, line[2], line[2]-0]) if file in added else table.append([line[0], 0, 0 - line[1], 0, 0-line[2]])
```

Reply:
Oh, I totally agree this part can be better written. The nested for loops is kinda bad. However, yours won't work since `table_old`, `table_new` are lists not dictionaries.

Reply:
oh my bad didnt realise it was a list...

yep i agree its slightly more readable now 👍 

is there any reason for not storing table as a dictionary? if you're going to loop through the list to find the relevant element, you might as well store a hash map with the dictionary.

Reply:
Just lookup the `tabulate` package and the print is doable with dictionary. I think it has been using list because it was easy and there wasn't a need for hash map yet.

Reply:
start to think is there actually a need for hash map? currently nested list works fine for what we need without making too many changes.